extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,335 @@
+import {assertHttpsUrl} from '../../../src/url';
+import {assert} from '../../../src/asserts';
+import {cidFor} from '../../../src/cid';
+import {getService} from '../../../src/service';
+import {log} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+
+const TAG_ = 'AmpUserNotification';
+
+/**
+ * @typedef {{
+ *   elementId: string,
+ *   ampUserId: string
+ * }}
+ */
+let GetRequestMetadata;
+
+/**
+ * @typedef {{
+ *   showNotification: boolean
+ * }}
+ */
+let GetResponseMetadata;
+
+/**
+ * @typedef {{
+ *   promise: (!Promise),
+ *   resolve: (function(*))
+ * }}
+ */
+let UserNotificationDefer;
+
+/**
+ * Component class that handles a simple notification.
+ */
+export class AmpUserNotification extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout != Layout.FILL;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!UserNotification} */
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    /** @private {!Promise} */
+    this.activatedPromise_ = null;
+
+    assert(this.element.hasAttribute('id'), 'amp-user-notification should ' +
+        'have an id.');
+
+    /** @private {string} */
+    this.elementId_ = this.element.getAttribute('id');
+
+    /** @private {?string} */
+    this.ampUserId_ = null;
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        'amp-user-notification should have `data-show-if-href` attribute.');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-show-if-href'), this.element);
+
+    assert(this.element.hasAttribute('data-dismiss-href'),
+        'amp-user-notification should have `data-dismiss-href` attribute.');
+
+    /** @private @const {string} */
+    this.dismissHref_ = assertHttpsUrl(
+        this.element.getAttribute('data-dismiss-href'), this.element);
+
+    this.options_ = {
+      credentials: this.element.getAttribute('credentials')
+    };
+
+    /** @private {?function(*)} */
+    this.layedoutResolve_ = null;
+
+    /** @private {?function(*)} */
+    this.dialogResolve_ = null;
+
+    /** @private @const {!Promise} */
+    this.layedoutPromise_ = new Promise((resolve, reject) => {
+      this.layedoutResolve_ = resolve;
+    });
+
+    /** @private @const {!Promise} */
+    this.dialogPromise_ = new Promise((resolve, reject) => {
+      this.dialogResolve_ = resolve;
+    });
+
+    this.userNotification_.registerUserNotification(
+        this.elementId_, this.layedoutResolve_, this.dialogPromise_);
+
+    this.registerAction('dismiss', this.dismiss.bind(this));
+  }
+
+  /** @override */
+  layoutCallback(visible) {
+    if (!this.activatedPromise_) {
+    // `layedoutPromise_` will block until there is no other
+    // `amp-user-notification` in display at this point in time.
+    this.activatedPromise_ = this.layedoutPromise_
```

Comment:
this current code waits, until the previous notification is dismissed until it fetches the next one its about to show. (through the blocked promise)